File: src/ufid.rs
// Unique file identifiers.
// We use a SHA-256 hash of the file's contents as its identifier.

use std::fs::File;
use std::io::{BufReader, Read};
use sha2::{Digest, Sha256};

// Unique File IDentifier
pub fn ufid(file_path: &str) -> Result<String, std::io::Error> {
    let file = File::open(file_path)?;
    let mut reader = BufReader::new(file);
    let mut hasher = Sha256::new();

    // Feed in 1024 bytes of the file's contents at a time to the hasher.
    let mut buffer = [0; 1024];
    loop {
        let count = reader.read(&mut buffer)?;
        if count == 0 {
            break;
        }
        hasher.update(&buffer[..count]);
    }

    let hash: sha2::digest::generic_array::GenericArray<u8, sha2::digest::typenum::UInt<sha2::digest::typenum::UInt<sha2::digest::typenum::UInt<sha2::digest::typenum::UInt<sha2::digest::typenum::UInt<sha2::digest::typenum::UInt<sha2::digest::typenum::UTerm, sha2::digest::consts::B1>, sha2::digest::consts::B0>, sha2::digest::consts::B0>, sha2::digest::consts::B0>, sha2::digest::consts::B0>, sha2::digest::consts::B0>> = hasher.finalize();
    Ok(format!("{:x}", hash))
}

File: src/ufid.py
# ufid.py
# Unique file identifiers

import hashlib
import os

KB = 1024
MB = 1024 * KB
GB = 1024 * GB


def ufid(file_path: str, chunk_size: int = MB) -> str:
    # `chunk_size` defaults to 1MB.
    # The hash is calculated from the filename prepended to the file contents.

    sha256_hash = hashlib.sha256()
    sha256_hash.update(os.path.basename(file_path).encode("utf-8"))  # Read in filename.
    with open(file_path, "rb") as file:
        for chunk in iter(lambda: file.read(chunk_size), b""):
            sha256_hash.update(chunk)  # Read in chunk contents.
        return sha256_hash.hexdigest()
